src/server/tablet_snap.rs

```diff
@@ -0,0 +1,526 @@
+pub fn send_snap(
+    env: Arc<Environment>,
+    mgr: TabletSnapManager,
+    security_mgr: Arc<SecurityManager>,
+    cfg: &Config,
+    addr: &str,
+    msg: RaftMessage,
+    limiter: Limiter,
+) -> Result<impl Future<Output = Result<SendStat>>> {
+    assert!(msg.get_message().has_snapshot());
+    let timer = Instant::now();
+    let send_timer = SEND_SNAP_HISTOGRAM.start_coarse_timer();
+    let key = TabletSnapKey::from_region_snap(
+        msg.get_region_id(),
+        msg.get_to_peer().get_id(),
+        msg.get_message().get_snapshot(),
+    );
+
+    let cb = ChannelBuilder::new(env)
+        .stream_initial_window_size(cfg.grpc_stream_initial_window_size.0 as i32)
+        .keepalive_time(cfg.grpc_keepalive_time.0)
+        .keepalive_timeout(cfg.grpc_keepalive_timeout.0)
+        .default_compression_algorithm(cfg.grpc_compression_algorithm())
+        .default_gzip_compression_level(cfg.grpc_gzip_compression_level)
+        .default_grpc_min_message_size_to_compress(cfg.grpc_min_message_size_to_compress);
+
+    let channel = security_mgr.connect(cb, addr);
+    let client = TikvClient::new(channel);
+    let (sink, receiver) = client.snapshot()?;
+    let send_task = async move {
+        let sink = sink.sink_map_err(Error::from);
+        let total_size = send_snap_files(&mgr, sink, msg, key.clone(), limiter).await?;
+        let recv_result = receiver.map_err(Error::from).await;
+        send_timer.observe_duration();
+        drop(client);
+        match recv_result {
+            Ok(_) => {
+                mgr.delete_snapshot(&key);
+                Ok(SendStat {
+                    key,
+                    total_size,
+                    elapsed: timer.saturating_elapsed(),
+                })
+            }
+            Err(e) => Err(e),
+        }
+    };
+    Ok(send_task)
+}
+
+async fn recv_snap_files(
+    snap_mgr: TabletSnapManager,
+    mut stream: impl Stream<Item = Result<SnapshotChunk>> + Unpin,
+    limit: Limiter,
+) -> Result<RecvTabletSnapContext> {
+    let head = stream.next().await.transpose()?;
+    let context = RecvTabletSnapContext::new(head)?;
+    let path = snap_mgr.get_tmp_name_for_recv(&context.key);
+    info!("begin to receive tablet snapshot files"; "file" => %path.display());
+    fs::create_dir_all(&path)?;
+    loop {
+        let mut chunk = match stream.next().await {
+            Some(Ok(mut c)) if !c.has_message() => c.take_data(),
+            Some(_) => {
+                return Err(box_err!("duplicated metadata"));
+            }
+            None => break,
+        };
+        // the format of chunk:
+        // |--name_len--|--name--|--content--|
+        let len = chunk[0] as usize;
+        let file_name = box_try!(std::str::from_utf8(&chunk[1..len + 1]));
+        let p = path.join(file_name);
+        let mut f = File::create(&p)?;
+        let mut size = chunk.len() - len - 1;
+        f.write_all(&chunk[len + 1..])?;
+        while chunk.len() >= SNAP_CHUNK_LEN {
```

Comment:
> while chunk.len() >= SNAP_CHUNK_LEN
It's not reliable to use the chunk.len() >=SNAP_CHUNK_LEN as the indicator that the previous file is ended or not. 
It's possible that the file's last part is exactly same as SNAP_CHUNK_LEN and then you'll merge the two files into one.